### PR TITLE
[SPARK-25595] Ignore corrupt Avro files if flag IGNORE_CORRUPT_FILES enabled

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

With flag `IGNORE_CORRUPT_FILES` enabled, schema inference should ignore corrupt Avro files, which is consistent with Parquet and Orc data source.

## How was this patch tested?

Unit test

### CODE REVIEW

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 1]
```diff
@@ -100,6 +77,50 @@ private[avro] class AvroFileFormat extends FileFormat
     }
   }
 
+  private def inferAvroSchemaFromFiles(
+      files: Seq[FileStatus],
+      conf: Configuration,
+      ignoreExtension: Boolean): Schema = {
+    val ignoreCorruptFiles = SQLConf.get.ignoreCorruptFiles
```
Comment (HyukjinKwon) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Bout about matching it to `sparkSession.sessionState.conf.ignoreCorruptFiles` like other occurrences?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 2]
```diff
@@ -100,6 +77,50 @@ private[avro] class AvroFileFormat extends FileFormat
...
+    // Schema evolution is not supported yet. Here we only pick first random readable sample file to
+    // figure out the schema of the whole dataset.
+    val avroReader = files.iterator.map { f =>
+      val path = f.getPath
+      if (!ignoreExtension && !path.getName.endsWith(".avro")) {
+        None
+      } else {
+        val in = new FsInput(path, conf)
```
Comment (HyukjinKwon) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Not a big deal but we can use `Utils.tryiWithResource`

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 3]
```diff
@@ -100,6 +77,50 @@ private[avro] class AvroFileFormat extends FileFormat
...
+          in.close()
+        }
+      }
+    }.collectFirst {
+      case Some(reader) => reader
+    }
+
+    avroReader match {
+      case Some(reader) =>
+        try {
```
Comment (HyukjinKwon) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: ditto for `tryWithResource`

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 4]
```diff
@@ -342,6 +342,53 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
     }
   }
 
+  private def createDummyCorruptFile(dir: File): Unit = {
+    FileUtils.forceMkdir(dir)
+    val corruptFile = new File(dir, "corrupt.avro")
+    val writer = new BufferedWriter(new FileWriter(corruptFile))
+    writer.write("corrupt")
+    writer.close()
```
Comment (HyukjinKwon) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: ditto for `tryWithResource`

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 5]
```diff
@@ -342,6 +342,53 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+
+        val srcFile = new File("src/test/resources/episodes.avro")
+        val destFile = new File(dir, "episodes.avro")
+        FileUtils.copyFile(srcFile, destFile)
+
+        val df = spark.read.format("avro").load(srcFile.getAbsolutePath)
+        val schema = df.schema
+        val result = df.collect()
+        // Schema inference picks random readable sample file.
+        // Here we use a loop to eliminate randomness.
```
Comment (HyukjinKwon) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: Actually I don't think it's randomness in this test. In this test, HDFS lists files in an alphabetical order under to the hood although it's not guaranteed. I think the picking order here at least is deterministic.

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 6]
```diff
@@ -342,6 +342,53 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+      }
+    }
+  }
+
+  test("Throws IOException on reading corrupt Avro file if flag IGNORE_CORRUPT_FILES disabled") {
+    withSQLConf(SQLConf.IGNORE_CORRUPT_FILES.key -> "false") {
+      withTempPath { dir =>
+        createDummyCorruptFile(dir)
+        val message = intercept[org.apache.spark.SparkException] {
+          spark.read.format("avro").load(dir.getAbsolutePath).schema
```
Comment (HyukjinKwon) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: `.schema` wouldn't probably be needed.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala (hunk 3 above)
Comment (gengliangwang) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: This one is too trivial.

**Comment by gengliangwang** [comment]
@HyukjinKwon Thanks for the review :)

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**